test/e2e/storage/drivers/csi.go

```diff
@@ -416,25 +598,97 @@ func (m *mockCSIDriver) PrepareTest(f *framework.Framework) (*storageframework.P
 		RequiresRepublish: m.requiresRepublish,
 		FSGroupPolicy:     m.fsGroupPolicy,
 	}
-	cleanup, err := utils.CreateFromManifests(f, driverNamespace, func(item interface{}) error {
+	cleanup, err := utils.CreateFromManifests(f, m.driverNamespace, func(item interface{}) error {
 		return utils.PatchCSIDeployment(f, o, item)
 	}, m.manifests...)
 
 	if err != nil {
 		framework.Failf("deploying csi mock driver: %v", err)
 	}
 
-	cleanupFunc := generateDriverCleanupFunc(
+	driverCleanupFunc := generateDriverCleanupFunc(
 		f,
 		"mock",
 		testns,
 		driverns,
 		cleanup,
 		cancelLogging)
 
+	cleanupFunc := func() {
+		embeddedCleanup()
+		driverCleanupFunc()
+	}
+
 	return config, cleanupFunc
 }
 
+func (m *mockCSIDriver) interceptGRPC(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+	defer func() {
+		// Always log the call and its final result,
+		// regardless whether the result was from the real
+		// implementation or a hook.
+		m.calls.LogGRPC(info.FullMethod, req, resp, err)
+	}()
+
+	if m.hooks.Pre != nil {
+		resp, err = m.hooks.Pre(ctx, info.FullMethod, req)
+		if resp != nil || err != nil {
+			return
+		}
+	}
+	resp, err = handler(ctx, req)
+	if m.hooks.Post != nil {
+		resp, err = m.hooks.Post(ctx, info.FullMethod, req, resp, err)
+	}
+	return
+}
+
+func (m *mockCSIDriver) GetCalls() ([]MockCSICall, error) {
```

Comment:
I still see some tests run with not embedded CSI drivers, is there plan to move them towards it? In such case a TODO comment to store intercepted calls above and returning them here would save us some ugly JSON parsing.

Reply:
Not for now. In some earlier revision of the PR I ran all tests with embedded mock driver and just captured calls directly in memory. But then @msau42 pointed out that tests that are meant to become part of conformance can't depend on the port forwarding API in Kubernetes because that itself currently isn't required for conformance.

My plan for now is to work a bit more with this new approach and get it exercised regularly as part of E2E testing to be confident with it.

Perhaps I should simply submit the existing port forwarding tests for inclusion in conformance testing.... I asked in https://kubernetes.slack.com/archives/C78F00H99/p1606322090151300 why it wasn't included but didn't get much feedback.